go/libkb/per_user_key.go

```diff
@@ -0,0 +1,598 @@
+func (s *PerUserKeyring) Update(ctx context.Context) (ret *PerUserKeyring, err error) {
+	ret, err = s.Clone()
+	if err != nil {
+		return nil, err
+	}
+	err = ret.Sync(ctx)
+	return ret, err
+}
+
+// Sync our PerUserKeyring with the server. It will either add all new
+// keys since our last update, or not at all if there was an error.
+// Pass it a standard Go network context.
+func (s *PerUserKeyring) Sync(ctx context.Context) (err error) {
+	return s.syncAsConfiguredDevice(ctx, nil, nil)
+}
+
+// `lctx` and `upak` are optional
+func (s *PerUserKeyring) SyncWithExtras(ctx context.Context, lctx LoginContext, upak *keybase1.UserPlusAllKeys) (err error) {
+	return s.syncAsConfiguredDevice(ctx, lctx, upak)
+}
+
+// `lctx` and `upak` are optional
+func (s *PerUserKeyring) syncAsConfiguredDevice(ctx context.Context, lctx LoginContext, upak *keybase1.UserPlusAllKeys) (err error) {
+	uid, deviceID, _, _, activeDecryptionKey := s.G().ActiveDevice.AllFields()
+	if !s.uid.Equal(uid) {
+		return fmt.Errorf("UID changed on PerUserKeyring")
+	}
+	if deviceID.IsNil() {
+		return fmt.Errorf("missing configured deviceID")
+	}
+	return s.sync(ctx, lctx, upak, deviceID, activeDecryptionKey)
+}
+
+// `lctx` and `upak` are optional
+func (s *PerUserKeyring) SyncAsPaperKey(ctx context.Context, lctx LoginContext, upak *keybase1.UserPlusAllKeys, deviceID keybase1.DeviceID, decryptionKey GenericKey) (err error) {
+	if deviceID.IsNil() {
+		return fmt.Errorf("missing deviceID")
+	}
+	// Note this `== nil` check might not work, as it might be a typed nil.
+	if decryptionKey == nil {
+		return fmt.Errorf("missing decryption key")
+	}
+	return s.sync(ctx, lctx, upak, deviceID, decryptionKey)
+}
+
+// `lctx` and `upak` are optional
+func (s *PerUserKeyring) sync(ctx context.Context, lctx LoginContext, upak *keybase1.UserPlusAllKeys, deviceID keybase1.DeviceID, decryptionKey GenericKey) (err error) {
+	defer s.G().CTrace(ctx, "PerUserKeyring#sync", func() error { return err })()
+
+	s.G().Log.CDebugf(ctx, "PerUserKeyring#sync(%v, %v)", lctx != nil, upak != nil)
+
+	s.Lock()
+	defer s.Unlock()
+
+	box, prevs, err := s.fetchBoxesLocked(ctx, lctx, deviceID)
+	if err != nil {
+		return err
+	}
+
+	if upak == nil {
+		upak, err = s.getUPAK(ctx, lctx, upak)
+		if err != nil {
+			return err
+		}
+	}
+
+	newKeys, err := s.importLocked(ctx, box, prevs, decryptionKey, newPerUserKeyChecker(upak))
+	if err != nil {
+		return err
+
+	}
+	s.mergeLocked(newKeys)
+	return nil
+}
+
+func (s *PerUserKeyring) getUPAK(ctx context.Context, lctx LoginContext, upak *keybase1.UserPlusAllKeys) (*keybase1.UserPlusAllKeys, error) {
+	if upak != nil {
+		return upak, nil
+	}
+	upakArg := NewLoadUserByUIDArg(ctx, s.G(), s.uid)
+	upakArg.LoginContext = lctx
+	upak, _, err := s.G().GetUPAKLoader().Load(upakArg)
+	return upak, err
+}
+
+func (s *PerUserKeyring) mergeLocked(m PerUserKeyMap) (err error) {
+	for k, v := range m {
+		s.generations[k] = v
```

Comment:
This is not a deep copy. But it points to an object that should never be mutated.
```
type perUserKeyFull struct {
	seed   PerUserKeySeed
	sigKey *NaclSigningKeyPair
	encKey *NaclDHKeyPair
}
```